refactor canonical ordering to also be able to do isAfter checks

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42141 [JIT] Make create autodiff subgraphs do in place updates to aliasDb
* **#42140 refactor canonical ordering to also be able to do isAfter checks**

Differential Revision: [D22798378](https://our.internmc.facebook.com/intern/diff/D22798378)

Commit:
refactor canonical ordering to also be able to do isAfter checks

[ghstack-poisoned]

Commit:
Update on "refactor canonical ordering to also be able to do isAfter checks"




[ghstack-poisoned]

Commit:
Update on "refactor canonical ordering to also be able to do isAfter checks"


Differential Revision: [D22798378](https://our.internmc.facebook.com/intern/diff/D22798378)

[ghstack-poisoned]

Commit:
Update on "refactor canonical ordering to also be able to do isAfter checks"


Differential Revision: [D22798378](https://our.internmc.facebook.com/intern/diff/D22798378)

[ghstack-poisoned]

Commit:
Update on "refactor canonical ordering to also be able to do isAfter checks"


Differential Revision: [D22798378](https://our.internmc.facebook.com/intern/diff/D22798378)

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/canonicalize.cpp:
```diff
@@ -129,20 +129,33 @@ bool isBefore(const Use& a, const Use& b) {
-      if (!isBefore(first_use, n_use)) {
-        first_use = n_use;
-      }
+bool isAfter(const Use& a, const Use& b) {
```
nitpick: should we move these to `ir.h`?

Krovatkin commented on torch/csrc/jit/passes/canonicalize.cpp:
```diff
@@ -129,20 +129,33 @@ bool isBefore(const Use& a, const Use& b) {
+  Use extreme_use = v->uses()[0];
+  for (size_t i = 1; i < v->uses().size(); ++i) {
+    auto n_use = v->uses()[i];
+    if (!isBeforeOrAfter(extreme_use, n_use, find_first)) {
```
nitpick: i don't if there's a lot of value in refactoring that ternary into a helper given there's just one use of it. it seems to make code a little bit harder to follow

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "refactor canonical ordering to also be able to do isAfter checks"


Differential Revision: [D22798378](https://our.internmc.facebook.com/intern/diff/D22798378)

[ghstack-poisoned]
